File: lib/services/stream_resolver.dart
import 'package:flutter/foundation.dart';
import 'dart:convert';
import 'package:http/http.dart' as http;

import '../services/playlist_service.dart';

/// A utility to automatically scrape and resolve live stream tokens
/// from web player pages.
class StreamResolver {
  /// Checks if a URL needs to be resolved (ends with |scrape=true or %7Cscrape=true)
  /// and returns the raw stream URL. Otherwise, returns the original URL.
  static Future<String> resolveIfNeeded(String url, {String? userAgent}) async {
    // 1. Check for stalker portal MAC URLs
    if (url.contains('mac=') && url.contains('live.php')) {
      try {
        var currentUrl = url;
        var redirects = 0;
        final client = http.Client();
        while (redirects < 5) {
          final request = http.Request('GET', Uri.parse(currentUrl))..followRedirects = false;
          request.headers['User-Agent'] = userAgent ?? 'SmartIPTV';
          final response = await client.send(request);
          if (response.statusCode >= 300 && response.statusCode < 400) {
            final location = response.headers['location'];
            if (location != null) {
              if (location.startsWith('http')) {
                currentUrl = location;
              } else if (location.startsWith('//')) {
                currentUrl = 'http:$location';
              } else if (location.startsWith('/')) {
                final uri = Uri.parse(currentUrl);
                currentUrl = '${uri.scheme}://${uri.host}${uri.hasPort ? ':${uri.port}' : ''}$location';
              } else {
                 final uri = Uri.parse(currentUrl);
                 final path = uri.path.substring(0, uri.path.lastIndexOf('/') + 1);
                 currentUrl = '${uri.scheme}://${uri.host}${uri.hasPort ? ':${uri.port}' : ''}$path$location';
              }
              redirects++;
              response.stream.drain();
            } else {
              response.stream.drain();
              break;
            }
          } else {
            response.stream.drain();
            break;
          }
        }
        client.close();
        if (currentUrl != url) {
          if (url.contains('extension=m3u8') && !currentUrl.contains('m3u8')) {
            return '$currentUrl#.m3u8';
          }
          return currentUrl;
        }
      } catch (e) {
        print('StreamResolver Stalker Error: $e');
      }
    }

    if (!url.endsWith('|scrape=true') && !url.endsWith('%7Cscrape=true')) {
      return url;
    }

    final decryptedUrl = Channel.decrypt(url);
    final targetUrl = decryptedUrl.replaceAll('|scrape=true', '').replaceAll('%7Cscrape=true', '');
    try {
      final res = await http.get(
        Uri.parse(targetUrl),
        headers: {
          'User-Agent': userAgent ?? 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36',
          'Accept': 'text/html,application/xhtml+xml,application/xml;q=0.9,image/avif,image/webp,image/apng,*/*;q=0.8',
        },
      );

      if (res.statusCode != 200) {
        return targetUrl; // Fallback
      }

      final body = res.body;

      // 1. Check for standard m3u8/mpd links in plain text
      final directMatch = RegExp(r'''(https?://[^\s'"]+\.(?:m3u8|mpd)[^\s'"]*)''').firstMatch(body);
      if (directMatch != null) {
        return directMatch.group(1)!;
      }

      // 2. Check for base64 encoded strings (like atob("..."))
      final atobRegExp = RegExp(r'''atob\(['"]([^'"]+)['"]\)''');
      final matches = atobRegExp.allMatches(body);
      for (final match in matches) {
        try {
          final b64 = match.group(1)!;
          final decoded = utf8.decode(base64.decode(b64));
          final streamMatch = RegExp(r'''(https?://[^\s'"]+\.(?:m3u8|mpd)[^\s'"]*)''').firstMatch(decoded);
          if (streamMatch != null) {
            return streamMatch.group(1)!;
          }
        } catch (_) {
          // Ignore decoding errors
        }
      }

      // 3. Check for iframe embeds
      final iframeRegExp = RegExp(r'''<iframe[^>]+src=['"]([^'"]+)['"]''', caseSensitive: false);
      final iframeMatch = iframeRegExp.firstMatch(body);
      if (iframeMatch != null) {
        var iframeSrc = iframeMatch.group(1)!;
        if (iframeSrc.startsWith('//')) {
          iframeSrc = 'https:$iframeSrc';
        } else if (iframeSrc.startsWith('/')) {
          final uri = Uri.parse(targetUrl);
          iframeSrc = '${uri.scheme}://${uri.host}$iframeSrc';
        }
        
        // Recursively scrape the iframe
        final iframeRes = await http.get(
          Uri.parse(iframeSrc),
          headers: {
            'User-Agent': userAgent ?? 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36',
            'Referer': targetUrl,
            'Accept': 'text/html,application/xhtml+xml,application/xml;q=0.9,image/avif,image/webp,image/apng,*/*;q=0.8',
          },
        );
        if (iframeRes.statusCode == 200) {
          final iframeBody = iframeRes.body;
          
          final iframeDirectMatch = RegExp(r'''(https?://[^\s'"]+\.(?:m3u8|mpd)[^\s'"]*)''').firstMatch(iframeBody);
          if (iframeDirectMatch != null) return iframeDirectMatch.group(1)!;
          
          final iframeMatches = atobRegExp.allMatches(iframeBody);
          for (final match in iframeMatches) {
            try {
              final b64 = match.group(1)!;
              final decoded = utf8.decode(base64.decode(b64));
              final streamMatch = RegExp(r'''(https?://[^\s'"]+\.(?:m3u8|mpd)[^\s'"]*)''').firstMatch(decoded);
              if (streamMatch != null) return streamMatch.group(1)!;
            } catch (e) { debugPrint('Caught error in stream_resolver.dart: $e'); }
          }
        }
      }

      // 4. Just general base64 matching (a bit risky, so restricted to typical lengths)
      // Usually not needed if atob or plain text works.

    } catch (e) {
      print('StreamResolver Error: $e');
    }

    return targetUrl; // Fallback to original
  }
}

File: android/app/src/main/kotlin/com/kobani4k/player/StreamResolver.kt
package com.kobani4k.player

import android.util.Base64
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import okhttp3.OkHttpClient
import okhttp3.Request
import java.util.regex.Pattern

object StreamResolver {
    private val client = OkHttpClient()
    private val ATOB_PATTERN = Pattern.compile("atob\\(['\"]([^'\"]+)['\"]\\)")
    private val STREAM_PATTERN = Pattern.compile("(https?://[^\\s'\"]+\\.(?:m3u8|mpd)[^\\s'\"]*)")

    suspend fun resolveIfNeeded(url: String, userAgent: String? = null): String {
        // 1. Check for stalker portal MAC URLs
        if (url.contains("mac=") && url.contains("live.php")) {
            return withContext(Dispatchers.IO) {
                var currentUrl = url
                var redirects = 0
                val ua = userAgent ?: "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36"
                try {
                    val noRedirectClient = client.newBuilder().followRedirects(false).followSslRedirects(false).build()
                    while (redirects < 5) {
                        val request = Request.Builder().url(currentUrl)
                            .header("User-Agent", ua)
                            .build()
                        val response = noRedirectClient.newCall(request).execute()
                        if (response.isRedirect) {
                            val location = response.header("Location")
                            if (location != null) {
                                if (location.startsWith("http")) {
                                    currentUrl = location
                                } else if (location.startsWith("//")) {
                                    currentUrl = "http:$location"
                                } else if (location.startsWith("/")) {
                                    val uri = java.net.URI(currentUrl)
                                    val portStr = if (uri.port != -1) ":${uri.port}" else ""
                                    currentUrl = "${uri.scheme}://${uri.host}$portStr$location"
                                } else {
                                    val uri = java.net.URI(currentUrl)
                                    val path = uri.path.substring(0, uri.path.lastIndexOf('/') + 1)
                                    val portStr = if (uri.port != -1) ":${uri.port}" else ""
                                    currentUrl = "${uri.scheme}://${uri.host}$portStr$path$location"
                                }
                                redirects++
                            } else {
                                break
                            }
                        } else {
                            break
                        }
                    }
                    if (currentUrl != url) {
                        if (url.contains("extension=m3u8") && !currentUrl.contains("m3u8")) {
                            return@withContext "$currentUrl#.m3u8"
                        }
                        return@withContext currentUrl
                    }
                } catch (e: Exception) {
                    // ignore
                }
                url
            }
        }

        if (!url.endsWith("|scrape=true") && !url.endsWith("%7Cscrape=true")) {
            return url
        }

        val targetUrl = url.replace("|scrape=true", "").replace("%7Cscrape=true", "")
        return withContext(Dispatchers.IO) {
            try {
                val requestBuilder = Request.Builder().url(targetUrl)
                val ua = userAgent ?: "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36"
                requestBuilder.header("User-Agent", ua)
                requestBuilder.header("Accept", "text/html,application/xhtml+xml,application/xml;q=0.9,image/avif,image/webp,*/*;q=0.8")

                val response = client.newCall(requestBuilder.build()).execute()
                val body = response.body?.string() ?: return@withContext targetUrl

                // 1. Check for plain text streams
                val directMatcher = STREAM_PATTERN.matcher(body)
                if (directMatcher.find()) {
                    return@withContext directMatcher.group(1) ?: targetUrl
                }

                // 2. Check for atob base64
                val atobMatcher = ATOB_PATTERN.matcher(body)
                while (atobMatcher.find()) {
                    try {
                        val b64 = atobMatcher.group(1)
                        if (b64 != null) {
                            val decoded = String(Base64.decode(b64, Base64.DEFAULT), Charsets.UTF_8)
                            val streamMatcher = STREAM_PATTERN.matcher(decoded)
                            if (streamMatcher.find()) {
                                return@withContext streamMatcher.group(1) ?: continue
                            }
                        }
                    } catch (e: Exception) {
                        // ignore
                    }
                }

                // 3. Check for iframe embeds
                val iframePattern = Pattern.compile("<iframe[^>]+src=['\"]([^'\"]+)['\"]", Pattern.CASE_INSENSITIVE)
                val iframeMatcher = iframePattern.matcher(body)
                if (iframeMatcher.find()) {
                    var iframeSrc = iframeMatcher.group(1) ?: return@withContext targetUrl
                    if (iframeSrc.startsWith("//")) {
                        iframeSrc = "https:$iframeSrc"
                    } else if (iframeSrc.startsWith("/")) {
                        val uri = java.net.URI(targetUrl)
                        iframeSrc = "${uri.scheme}://${uri.host}$iframeSrc"
                    }

                    try {
                        val iframeRequest = Request.Builder().url(iframeSrc)
                            .header("User-Agent", ua)
                            .header("Referer", targetUrl)
                            .header("Accept", "text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8")
                            .build()
                        val iframeResponse = client.newCall(iframeRequest).execute()
                        val iframeBody = iframeResponse.body?.string() ?: return@withContext targetUrl

                        val iframeDirectMatcher = STREAM_PATTERN.matcher(iframeBody)
                        if (iframeDirectMatcher.find()) return@withContext iframeDirectMatcher.group(1) ?: targetUrl

                        val iframeAtobMatcher = ATOB_PATTERN.matcher(iframeBody)
                        while (iframeAtobMatcher.find()) {
                            try {
                                val b64 = iframeAtobMatcher.group(1)
                                if (b64 != null) {
                                    val decoded = String(Base64.decode(b64, Base64.DEFAULT), Charsets.UTF_8)
                                    val streamMatcher = STREAM_PATTERN.matcher(decoded)
                                    if (streamMatcher.find()) return@withContext streamMatcher.group(1) ?: continue
                                }
                            } catch (e: Exception) {}
                        }
                    } catch (e: Exception) {}
                }
                
                targetUrl
            } catch (e: Exception) {
                targetUrl
            }
        }
    }
}
